[Mangler] Handle local archetypes in `getDeclTypeForMangling`

hamishknight opened the pull request:
Element archetypes can occur here when mangling the USR for local variables for e.g SourceKit cursor info, as well as for regular compilation for things like lazy variables.

Update `getDeclTypeForMangling` to map local archetypes out of context, using both the captured generic environments and the archetypes present in the type. More work is needed to support lazy variable though (now it crashes in SILGen).

This patch doesn't handle mangling standalone element archetypes for e.g `printTypeUSR`, ideally we'd fix the clients there to not pass local archetypes.

rdar://145737847

hamishknight commented:
@swift-ci please test

slavapestov commented on lib/AST/ASTMangler.cpp:
```diff
@@ -2372,6 +2376,20 @@ void ASTMangler::appendImplFunctionType(SILFunctionType *fn,
+    captureInfo = fn->getCaptureInfo();
+
+  auto interfaceTy = mapLocalArchetypesOutOfContext(
+      archetype, sig, captureInfo.getGenericEnvironments());
```
Why not just call mapLocalArchetypesOutOfContext() before calling into the mangler?

hamishknight replied:
Not sure if this is what you're asking for, but could move this to `getDeclTypeForMangling` instead? I don't really have any strong opinions one way or the other

hamishknight replied:
Updated to move into `getDeclTypeForMangling`, how does this look @slavapestov?

slavapestov commented:
I'm afraid this is not the correct fix. It sounds like there is an issue with `lazy` variables, and something separate with CursorInfo.

slavapestov commented:
For `lazy` vars it appears you need to change `appendAccessorEntity()` to do the same thing as `appendClosureComponents()` when the accessor in question is a local function.

hamishknight commented:
> It sounds like there is an issue with `lazy` variables, and something separate with CursorInfo

It's the same underlying issue; we can't mangle a decl with an interface type that contains a local archetype

> Neither of these should require anything resembling a complete rethink of local archetypes to resolve.

Right, I was referring to handling standalone local archetypes for e.g `printTypeUSR`

> For `lazy` vars it appears you need to change `appendAccessorEntity()` to do the same thing as `appendClosureComponents()` when the accessor in question is a local function.

Is there any advantage to doing this over handling it generally for any decl?

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

slavapestov commented on lib/AST/LocalArchetypeRequirementCollector.cpp:
```diff
@@ -183,13 +183,29 @@ Type MapLocalArchetypesOutOfContext::getInterfaceType(
+  // If the environment we have is not for a capture, this must be a generic
+  // environment local to the parent function. For element archetypes, the depth
+  // we have is relative to the parent signature, so we can compute the correct
+  // depth by offsetting it by the number of parent captured environments.
```
This might not be correct if you have multiple nested `for` loops over packs. What causes us to end up in this code path?

hamishknight replied:
This handles e.g:

```swift
func foo<each T>(_ xs: repeat each T) {
  for x in repeat each xs {}
}
```

There is no captured environment for `x`, it's local to the function itself. Multiple nested should be fine since we track the parent generic signature in the statement checker, let me update the test case

hamishknight replied:
Oh hrm, looks like it doesn't work in a closure 🤔 Seems like we don't pass down the innermost element environment

hamishknight replied:
That isn't really an issue with this patch though, this crashes on main:

```swift
func foo<each T>(xs: repeat each T) {
  _ = {
    for x in repeat each xs {
      for y in repeat each xs {
        return (x, y)
      }
    }
  }
}
```

hamishknight commented:
@swift-ci please test

hamishknight committed:
[Mangler] Handle local archetypes in `getDeclTypeForMangling`

Element archetypes can occur here when mangling the USR for local
variables for e.g SourceKit cursor info, as well as for regular
compilation for things like lazy variables.

Update `getDeclTypeForMangling` to map local archetypes out of
context, using both the captured generic environments and the
archetypes present in the type. More work is needed to support lazy
variable though (now it crashes in SILGen).

This patch doesn't handle mangling standalone element archetypes for
e.g `printTypeUSR`, ideally we'd fix the clients there to not pass
local archetypes.

rdar://143077965

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/AST/ASTMangler.cpp:
```diff
@@ -4016,6 +4016,43 @@ static bool isMethodDecl(const Decl *decl) {
+  if (auto fn = AnyFunctionRef::fromDeclContext(innerDC))
+    captureInfo = fn->getCaptureInfo();
+
+  // Record any captured generic environments we have.
+  llvm::SmallSetVector<GenericEnvironment *, 4> capturedEnvs;
+  for (auto *genericEnv : captureInfo.getGenericEnvironments())
+    capturedEnvs.insert(genericEnv);
```
In principle I think we could purely rely on gathering them from the type, but I don't think there's any harm in doing this and it's more consistent with the closure logic. I don't feel strongly about it though.

hamishknight commented:
@slavapestov How does this look?
